### ISSUE TITLE
Resolution Loop 3: "Seek and destroy"

### ISSUE THREAD

**Issue by pawamoy** [pdm-project/pdm#781]
- [x] I have searched the issue tracker and believe that this is not a duplicate.

The next chapter of the saga!

Related, resolved issues: #627, #744, #776.

## Steps to reproduce

```toml
[build-system]
requires = ["pdm-pep517"]
build-backend = "pdm.pep517.api"

[project]
name = "ml-project"
description = "ML stuff"
authors = [{name = "someone", email = "someone@hey.com"}]
readme = "README.md"
requires-python = ">=3.6.8"
dynamic = ["version"]
dependencies = [
    "sentence-transformers==2.0.0",
    "torchvision==0.9.1",
    "torch==1.9.1",
    "importlib-metadata>=2.1.0",
]
```

```bash
pdm lock
```

## Actual behavior

Upon locking, PDM will loop on multiple versions of `importlib-metadata`, all the way down to version 2.0.0 or something, and finally print the actual issue:

```
✖ 🔒 Lock failed
Unable to find a resolution for torch==1.9.1 because the following requirements conflict:
  torch==1.8.1 (from <Candidate torchvision 0.9.1 from https://index/simple/torchvision/>)
  torch==1.9.1 (from project)
  torch>=1.6.0 (from <Candidate sentence-transformers 2.0.0 from https://index/simple/sentence-transformers/>)
To fix this, you could try to loosen the dependency version constraintsin pyproject.toml.
See /tmp/pdm-lock-hcodkw1j.log for detailed debug log.
[ResolutionImpossible]: Unable to find a resolution
Add '-v' to see the detailed traceback
```

I believe the packages on which it loops is arbitrary/random (probably just a matter of order). For example, if I add pandas:

```toml
dependencies = [
    "sentence-transformers==2.0.0",
    "torchvision==0.9.1",
    "torch==1.9.1",
    "importlib-metadata>=2.1.0",
    "pandas==1.1.4",
]
```

...PDM will loop on `python-dateutil` instead, this time for real (infinite loop: 2.8.2 to 2.7.3 then back to 2.8.2).

If I reduce the set of dependencies to torch and torchvision (and optionally sentence-transformers), PDM is able to directly print the error shown above.

## Expected behavior

When a compatibility issue is found, don't iterate on irrelevant packages and immediately stop resolution and print the error (this is probably a naive view of the problem/solution).

## Environment Information

```console
% pdm info && pdm info --env
PDM version:        1.11.1
Python Interpreter: /home/user/.basher-packages/pyenv/pyenv/versions/3.8.11/bin/python3 (3.8)
Project Root:       /home/user/data/dev/project
Project Packages:   /home/user/data/dev/project/__pypackages__/3.8
{
  "implementation_name": "cpython",
  "implementation_version": "3.8.11",
  "os_name": "posix",
  "platform_machine": "x86_64",
  "platform_release": "3.10.0-862.6.3.el7.x86_64",
  "platform_system": "Linux",
  "platform_version": "#1 SMP Fri Jun 15 17:57:37 EDT 2018",
  "python_full_version": "3.8.11",
  "platform_python_implementation": "CPython",
  "python_version": "3.8",
  "sys_platform": "linux"
}
```

P.S.: feel free to edit this dumb title to better represent the actual issue 😉

**Comment by frostming** [comment]
Can you try the fix in #782, also some regression testing may be good

**Comment by pawamoy** [comment]
Damn, something is broken in my setup. The SSL read method blocks indefinitely, `pdm lock` hangs:

```python
  ...
  File "/home/pawamoy/.local/pipx/venvs/pdm/lib/python3.9/site-packages/pdm/models/candidates.py", line 248, in prepare
    downloaded = pip_shims.unpack_url(
  File "/home/pawamoy/.local/pipx/shared/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 213, in unpack_url
    file = get_http_url(
  File "/home/pawamoy/.local/pipx/shared/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 94, in get_http_url
    from_path, content_type = download(link, temp_dir.path)
  File "/home/pawamoy/.local/pipx/shared/lib/python3.9/site-packages/pip/_internal/network/download.py", line 145, in __call__
    for chunk in chunks:
  File "/home/pawamoy/.local/pipx/shared/lib/python3.9/site-packages/pip/_internal/network/utils.py", line 63, in response_chunks
    for chunk in response.raw.stream(
  File "/home/pawamoy/.local/pipx/shared/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 576, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/home/pawamoy/.local/pipx/shared/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 519, in read
    data = self._fp.read(amt) if not fp_closed else b""
  File "/home/pawamoy/.local/pipx/shared/lib/python3.9/site-packages/pip/_vendor/cachecontrol/filewrapper.py", line 62, in read
    data = self.__fp.read(amt)
  File "/home/pawamoy/.basher-packages/pyenv/pyenv/versions/3.9.9/lib/python3.9/http/client.py", line 463, in read
    n = self.readinto(b)
  File "/home/pawamoy/.basher-packages/pyenv/pyenv/versions/3.9.9/lib/python3.9/http/client.py", line 507, in readinto
    n = self.fp.readinto(b)
  File "/home/pawamoy/.basher-packages/pyenv/pyenv/versions/3.9.9/lib/python3.9/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
  File "/home/pawamoy/.basher-packages/pyenv/pyenv/versions/3.9.9/lib/python3.9/ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "/home/pawamoy/.basher-packages/pyenv/pyenv/versions/3.9.9/lib/python3.9/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
KeyboardInterrupt
```
